src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Generated.cs

```diff
@@ -12,6 +14,24 @@ partial class LambdaCompiler
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1505:AvoidUnmaintainableCode"), System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1502:AvoidExcessiveComplexity")]
         private void EmitExpression(Expression node, CompilationFlags flags)
         {
+            // When compling deep trees, we run the risk of triggering a terminating StackOverflowException,
+            // so we probe for sufficient stack here. When we run out of stack, we simply continue the work
+            // on another thread. Note that EnsureSufficientExecutionStack is very fast, so the overhead of
+            // this overflow protection is minimal.
```

Comment:
Another concern is that if we have possibilities for infinitely recursive cases, we would be trading crashes for OOMs and on 64bit OOMs can take forever. Sometimes quick crash is more preferrable than excruciating resource consuming OOM.

I just hope that infinite recursion is not a case for us here. I.E. - expression nodes are immutable and thus cannot be made self contained by an external builder. As such you cannot construct an infinitely large tree. 
Is that a valid assumption?